PR: fix PDF loader when using default Content Extraction Engine and OCR

rgaricano opened the pull request:
# Pull Request
## Fix PDF loader when using default Content Extraction Engine and OCR


## Checklist
- [ x] **Target branch:** Please verify that the pull request targets the `dev` branch.
- [ x] **Description:** Provide a concise description of the changes made in this pull request.
- [ ] **Changelog:** Ensure a changelog entry following the format of [Keep a Changelog](https://keepachangelog.com/) is added at the bottom of the PR description.
- [ ] **Documentation:** Have you updated relevant documentation [Open WebUI Docs](https://github.com/open-webui/docs), or other documentation sources?
- [x ] **Dependencies:** Are there any new dependencies? Have you updated the dependency versions in the documentation?
- [ x] **Testing:** Have you written and run sufficient tests for validating the changes?
- [ x] **Code review:** Have you performed a self-review of your code, addressing any coding standard issues and ensuring adherence to the project's coding standards?
- [ ] **Prefix:** To cleary categorize this pull request, prefix the pull request title, using one of the following:
- **BREAKING CHANGE**: Significant changes that may affect compatibility
- **build**: Changes that affect the build system or external dependencies
  - **ci**: Changes to our continuous integration processes or workflows
  - **chore**: Refactor, cleanup, or other non-functional code changes
  - **docs**: Documentation update or addition
  - **feat**: Introduces a new feature or enhancement to the codebase
  - **fix**: Bug fix or error correction
  - **i18n**: Internationalization or localization changes
  - **perf**: Performance improvement
- **refactor**: Code restructuring for better maintainability, readability, or scalability
- **style**: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc.)
  - **test**: Adding missing tests or correcting existing tests
- **WIP**: Work in progress, a temporary label for incomplete or ongoing work

# Changelog Entry
Fix PDF loader when using default CEE and OCR

### Description
Fix PDF loader when using default Content Extraction Engine and OCR enabled

Commit to fixing issues due to incorrect dimension calculation when reshaping images using the Langchain class Pypdfloader.

Using the UnsestructuredPDFLoader call instead of the PyPDFLoader call fixes errors encountered when processing PDF files with images.

(before I tried with PyMuPDF, is fast and do the work well, but with AGPL License, it was't approved,
In this case I use unstructured lib, same as is allready been used, with apache license.
unstructured[pdf] is an extra option with that requirements:
Provides-Extra: pdf
Requires-Dist: onnx ; extra == 'pdf'
Requires-Dist: pdf2image ; extra == 'pdf'
Requires-Dist: pdfminer.six ; extra == 'pdf'
Requires-Dist: pikepdf ; extra == 'pdf'
Requires-Dist: pi-heif ; extra == 'pdf'
Requires-Dist: pypdf ; extra == 'pdf'
Requires-Dist: google-cloud-vision ; extra == 'pdf'
Requires-Dist: effdet ; extra == 'pdf'
Requires-Dist: unstructured-inference (>=0.8.7) ; extra == 'pdf'
Requires-Dist: unstructured.pytesseract (>=0.3.12) ; extra == 'pdf'


As I don't know if all unstructured options are going to be installed, i added a specific entry in requirements.txt: unstructured[pdf]

Note: when testing it give this errors:
- No module named 'unstructured_inference'
	(I run `pip install unstructured_inference`)
& after
-  No module named 'pdf2image'
	(I run `pip install pdf2image`)

Please check that, maybe some of these dependencies may not be desired. ?
unstructured_inference requirements
-------------------
Requires-Dist: python-multipart
Requires-Dist: huggingface-hub
Requires-Dist: numpy (<2)
Requires-Dist: opencv-python (!=4.7.0.68)
Requires-Dist: onnx
Requires-Dist: onnxruntime (>=1.17.0)
Requires-Dist: matplotlib
Requires-Dist: torch
Requires-Dist: timm
Requires-Dist: transformers (>=4.25.1)
Requires-Dist: rapidfuzz
Requires-Dist: pandas
Requires-Dist: scipy
Requires-Dist: pypdfium2
Requires-Dist: pdfminer-six (==20240706)

With this steps seem it work ok,
In this question you may need to do some testing to make sure all dependencies are installed correctly, if not you may need to explicitly list those libraries.

Licenses that I revised (pdfminer & pdf2image), others than own unstructured, are MIT.

### Added

### Changed
replace call to langchain-community PyPDFLoader class by langchain-community UnstructuredPDFLoader class

### Deprecated

### Removed

### Fixed
Errors uploading pdf files with default content extraction engine (internal) and OCR feature

### Security

### Breaking Changes
---

### Additional Information
https://github.com/open-webui/open-webui/discussions/11171
https://github.com/open-webui/open-webui/discussions/4458

### Screenshots or Videos
![openwebui_pdf_default_unstructured_ocrs_relevant](https://github.com/user-attachments/assets/a8556afc-c0fd-4053-911f-85f29c934a3b)
![openwebui_pdf_default_unstructured_ocrs_complete](https://github.com/user-attachments/assets/b9d44e32-f6e5-4a21-a955-83cb01fc62f5)
![openwebui_pdf_UED](https://github.com/user-attachments/assets/cc515210-2552-4ac7-a69a-1722a155dbce)
![openwebui_pdf_UFR](https://github.com/user-attachments/assets/961911ac-fc8b-489f-a457-3928201759e5)

tjbck committed:
refac: assets

rgaricano committed:
Fix PDF loader when using default CEE and OCR

Commit to fix issues due a bad calculate of dimmensions when reshape
images by langchain lib pypdfloader.

It was fixes on langchain using pymupdf lib, that do this task better,
in replace of pypdf.

rgaricano committed:
Merge branch 'dev' of https://github.com/rgaricano/open-webui into dev

rgaricano committed:
Update main.py

removing some wrong chars typed by error

rgaricano committed:
Merge branch 'open-webui:dev' into dev

rgaricano committed:
Update main.py fix PDF loader when using default Content Extraction Engine and OCR

fix PDF loader when using default Content Extraction Engine and OCR

Replaced LangChain-community class PypPDFLoader by UnstructuredPDFLoader

rgaricano committed:
Update requirements.txt for fix PDF default engine+OCR

removed library pypdf dependency in favour of unstructured[pdf]

rgaricano committed:
Update pyproject.toml 

Update dependencis for fix PDF default engine with OCR lib sustitution (pypdfloader by UnstructuredPDFLoader)

rgaricano commented:
what kind of performances, time & database space? accurace of ocr & text extraction? accurace in responses after vectorized proccess? with different embedd models?

I don't usually do many performance tests, other than my own eye on the performance and fluidity of the system.
I'm a senior and in some aspects a bit rusty already.
But I can try or delegate it to some user interested in collaborating.
(The drawback is that with a library that fails, few comparative tests can be done!)

tjbck commented:
I just want to understand the tradeoff for switching to unstructured, just overall speed and edge cases comparisons would be helpful here!

rgaricano committed:
Merge branch 'open-webui:dev' into dev

rgaricano commented:
I did some tests,

rgaricano committed:
Merge branch 'open-webui:dev' into dev

dliebrecht commented:
I am new to GitHub.  I am waiting for this fix.  Hoping this GROK output is helpful.  Apologies if this isn't appropriate to post.  Let me know.

Key Tradeoffs Between PyPDFLoader and UnstructuredPDFLoader
The tradeoff primarily revolves around balancing speed, capability, and accuracy, as well as considering dependencies and resource demands. Below is a breakdown of what this means:

Speed
PyPDFLoader: This loader is generally faster for simple, text-based PDFs because it uses PyPDF2, a lightweight, pure-Python library focused on basic text extraction. It performs minimal processing, which keeps it quick for straightforward documents.
UnstructuredPDFLoader: This loader may be slower because it performs more advanced processing, such as OCR (Optical Character Recognition) and layout analysis, especially when handling PDFs with images or complex structures. The additional computational overhead comes from its reliance on the Unstructured library, which is designed for more robust document parsing.
Capability
PyPDFLoader: It struggles with complex PDFs, particularly those with images, tables, or non-standard layouts. In the case of Open WebUI, it produced errors when processing PDFs with images under the default CEE and OCR settings, prompting the need for this PR.
UnstructuredPDFLoader: This loader excels at handling a wider range of PDFs, including scanned documents, images, and intricate layouts. It can extract text from images via OCR and analyze document structure, making it more versatile and reliable for diverse user-uploaded PDFs in Open WebUI.
Accuracy
PyPDFLoader: Its text extraction accuracy is lower for PDFs with images or scanned content because it lacks OCR capabilities. This leads to incomplete or incorrect data, which can degrade the quality of responses after the extracted text is vectorized for use in Open WebUI’s AI features (e.g., Retrieval Augmented Generation, or RAG).
UnstructuredPDFLoader: By incorporating OCR and advanced text extraction, it provides higher accuracy for complex PDFs. Better text extraction translates to more precise vectorized data and, consequently, more accurate AI responses, especially for documents with images or non-text elements.
Dependencies and Resource Demands
PyPDFLoader: Requires fewer dependencies (mainly PyPDF2), making it lighter and easier to integrate into projects with minimal setup.
UnstructuredPDFLoader: Comes with a broader set of dependencies, included via unstructured[pdf] in the PR’s updated requirements.txt. These include libraries like pdf2image, unstructured_inference, and others for OCR and image processing. While this increases the setup complexity and resource usage (e.g., memory and CPU), it’s necessary for its enhanced functionality.
Why the Tradeoff Matters in Open WebUI
Open WebUI is a user-friendly platform where users can upload various PDFs for processing with AI features like RAG. These PDFs might range from simple text documents to scanned files with images or tables. The switch to UnstructuredPDFLoader ensures broader compatibility and fixes the errors encountered with PyPDFLoader, but it introduces the possibility of slower processing times. Tjbck’s request for a "performance difference table with tradeoffs" aims to quantify this—essentially asking, "Is the improved capability worth the potential speed cost?"

Conclusion
The tradeoff with using UnstructuredPDFLoader instead of PyPDFLoader means accepting potentially slower processing times due to its advanced features (like OCR and layout analysis), in exchange for better capability and accuracy when handling complex PDFs with images or non-standard layouts. For Open WebUI, where users expect reliable handling of diverse documents, this switch is a practical improvement, even though it may demand more computational resources. To fully address tjbck’s concern, benchmarks comparing speed and edge-case performance (e.g., simple vs. image-heavy PDFs) would help justify the change.

bet0x commented:
unstructured was good til they made their model for OCR paid, so [pyfidum](https://github.com/pypdfium2-team/pypdfium2) or any of those beats the free version. Adding such requirement should be optional, I'm personally replacing my old unstructured code with Docling.

rgaricano committed:
Merge branch 'open-webui:dev' into dev

rgaricano committed:
Fix Download pdf with MD formatting

Fix Download pdf with MD formatting

Replace unused markdown lib with fpdf2 recommended [mistletoe lib](https://github.com/miyuchina/mistletoe] 
(MIT license)

rgaricano committed:
FIX Download pdf with MD - Update requirements.txt

FIX Download pdf with MD - Update requirements.txt
 replaze markdown lib by mistletoe, details on pdf_generator.py commit

rgaricano committed:
Fix Download pdf with MD - update pyproject.toml

Fix Download pdf with MD - update pyproject.toml

Update pyproject.toml for update markdown library, replaced by mistletoe, for fix pdf generation.

biishan commented:
@rgaricano Much appreciated the contribution. Can this be merged into dev ?
